cpp/src/arrow/compute/kernels/aggregation.h

```diff
@@ -0,0 +1,84 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_COMPUTE_KERNELS_AGGREGATION_H
+#define ARROW_COMPUTE_KERNELS_AGGREGATION_H
+
+#include <memory>
+#include <mutex>
+
+#include "arrow/compute/kernel.h"
+
+namespace arrow {
+
+class Array;
+class Status;
+
+namespace compute {
+
+struct Datum;
+class FunctionContext;
+
+/// \class AggregateState
+/// \brief Interface for aggregate kernels.
+///
+/// An AggregateState separate the concerns of kernel computation and parallel
+/// scheduling.
+///
+/// \code{.cpp}
+/// // Loop can run in parallel.
```

Comment:
I think this is only true for Monoids?

Reply:
Do you have an example of an aggregate function which is not associative?

Reply:
Short answer (might be contrived): "exponentially smoothed moving average" and "hash elements"
Longer answer (probably obvious and too pedantic):   I was assuming a naive parallelization where Consume could be called in any order which would require Commutativity and associativity. With non-commutative associative operations (e.g. string concatenation, FIRST, LAST, etc ) there are more book-keeping details to to ensure parallelization still returns the correct results.  At the moment I'm having a hard time coming up with an example that benefits from parallelization of onsume that is not commutative.    

From a practical standpoint for this review:
*It might be good is add some additional pseudo-code for the parallelism AggregateState is meant to handle, * Possibly having some mechanism to signal commutative and non-commutative state.

Reply:
To answer my own question about associate and not commutative multithreading, the parsing of CSV files is a good example where multithreading makes sense.